File: leetcode/wordPattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function (pattern, s) {
  let hash = new Map();
  let sArray = s.split(' ');
  let seen = new Set();
  for (let i = 0; i <= pattern.length; i++) {
    let before = hash.get(pattern[i]);
    if (before) {
      if (before !== sArray[i]) return false;
    } else if (seen.has(sArray[i])) {
      return false;
    } else {
      hash.set(pattern[i], sArray[i]);
      seen.add(sArray[i]);
    }
  }
  return true;
};
wordPattern('abba', 'dog cat cat fish');
// class Solution {
//   public boolean wordPattern(String pattern, String s) {
//       HashMap map_index = new HashMap();
//       String[] words = s.split(" ");

//       if (words.length != pattern.length())
//           return false;

//       for (Integer i = 0; i < words.length; i++) {
//           char c = pattern.charAt(i);
//           String w = words[i];

//           if (!map_index.containsKey(c))
//               map_index.put(c, i);

//           if (!map_index.containsKey(w))
//               map_index.put(w, i);

//           if (map_index.get(c) != map_index.get(w))
//               return false;
//       }

//       return true;
//   }
// }

File: leetcode/wordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        return word_patter_2_pointer(pattern, s)


def word_patter_2_pointer(pattern, str: str):
    words = str.split(' ')
    n1, n2 = len(pattern), len(words)
    d = dict()
    s = set()
    i = 0
    while i < n1 and i < n2:
        if d[pattern[i]] in d:
            if d[pattern[i]] != words[i]:
                return False
        elif words[i] in s:
            return False
        else:
            d[pattern[i]] = words[i]
            s.add(words[i])
    return True
